libraries/eosiolib/contracts/eosio/key_value.hpp

```diff
@@ -73,7 +72,46 @@ namespace eosio {
    }
 
 namespace detail {
-   constexpr inline size_t max_stack_buffer_size = 512;
+   template <typename R, typename C>
+   auto member_pointer_ret_type(R (C::*)) -> R;
+   template <typename R, typename C, typename... Args>
+   auto member_pointer_ret_type(R (C::*)(Args...)) -> R;
+   template <typename R, typename C, typename... Args>
+   auto member_pointer_ret_type(R (C::*)(Args...)const) -> R;
+
+   template <auto MP>
+   using member_pointer_ret_t = decltype(member_pointer_ret_type(MP));
+
+   struct scoped_buffer {
+      constexpr static inline size_t static_size = 512;
+      constexpr inline scoped_buffer(std::size_t sz)
+         : ptr(sz > static_size ? new char[sz] : buffer.data()),
+           sz(sz) {}
+
+      ~scoped_buffer() {
+         if (size() > static_size)
+            delete[] ptr;
+      }
+
+      constexpr inline void reset(std::size_t new_size) {
+         if (new_size > size()) {
+            if (new_size > static_size) {
+               if (size() > static_size)
+                  delete[] ptr;
+               ptr = new char[new_size];
+            }
+         }
+         sz = new_size;
```

Comment:
It was doing that by design, the size of it is always guaranteed to be greater than the new_size on that leg of the branch (I didn't want to unduly allocate if I didn't have to.  There is a bit of an issue here if you 'reset' more than once though as we don't capture the 'original' size for deallocation purposes.  But, as this will be a temporary measure and be completely replaced after our next release, I am fine with it as is.

Reply:
Thanks for the explanation!